File: ts/src/data/stores.ts
import { existsSync, readFileSync, statSync } from "node:fs";
import { isAbsolute, relative, resolve } from "node:path";
import AdmZip from "adm-zip";

export interface JsonStore {
  exists(path: string): boolean;
  readText(path: string): string;
  readJson<T = unknown>(path: string): T;
  describe(): string;
}

function normalizePath(path: string): string {
  const normalized = path.replace(/\\/g, "/");
  if (normalized.startsWith("/")) {
    throw new Error(`Unsafe dataset path: ${path}`);
  }
  const parts = normalized.split("/").filter((part) => part !== "" && part !== ".");
  if (parts.includes("..")) {
    throw new Error(`Unsafe dataset path: ${path}`);
  }
  return parts.join("/");
}

function isInsideRoot(root: string, target: string): boolean {
  const rel = relative(root, target);
  return rel === "" || (!rel.startsWith("..") && !isAbsolute(rel));
}

export class DirectoryStore implements JsonStore {
  readonly root: string;

  constructor(root: string) {
    this.root = root;
  }

  private resolvePath(path: string): string {
    const root = resolve(this.root);
    const target = resolve(this.root, normalizePath(path));
    if (!isInsideRoot(root, target)) {
      throw new Error(`Unsafe dataset path: ${path}`);
    }
    return target;
  }

  resolveForDiagnostics(path: string): string {
    return this.resolvePath(path);
  }

  exists(path: string): boolean {
    const target = this.resolvePath(path);
    return existsSync(target) && statSync(target).isFile();
  }

  readText(path: string): string {
    const target = this.resolvePath(path);
    if (!existsSync(target) || !statSync(target).isFile()) {
      throw new Error(`Dataset file not found: ${path}`);
    }
    return readFileSync(target, "utf-8");
  }

  readJson<T = unknown>(path: string): T {
    return JSON.parse(this.readText(path)) as T;
  }

  describe(): string {
    return `directory:${this.root}`;
  }
}

export class ZipStore implements JsonStore {
  readonly zipPath: string;
  private _zip: AdmZip | null = null;

  constructor(zipPath: string) {
    this.zipPath = zipPath;
  }

  private zip(): AdmZip {
    if (this._zip === null) {
      this._zip = new AdmZip(this.zipPath);
    }
    return this._zip;
  }

  exists(path: string): boolean {
    const innerPath = normalizePath(path);
    if (!existsSync(this.zipPath)) return false;
    return this.zip().getEntry(innerPath) !== null;
  }

  readText(path: string): string {
    const innerPath = normalizePath(path);
    const entry = this.zip().getEntry(innerPath);
    if (!entry) {
      throw new Error(`Dataset zip entry not found: ${path}`);
    }
    return entry.getData().toString("utf-8");
  }

  readJson<T = unknown>(path: string): T {
    return JSON.parse(this.readText(path)) as T;
  }

  describe(): string {
    return `zip:${this.zipPath}`;
  }
}

export class FallbackStore implements JsonStore {
  readonly primary: JsonStore;
  readonly fallback: JsonStore;

  constructor(primary: JsonStore, fallback: JsonStore) {
    this.primary = primary;
    this.fallback = fallback;
  }

  private storeFor(path: string): JsonStore | null {
    if (this.primary.exists(path)) return this.primary;
    if (this.fallback.exists(path)) return this.fallback;
    return null;
  }

  exists(path: string): boolean {
    return this.storeFor(path) !== null;
  }

  readText(path: string): string {
    const store = this.storeFor(path);
    if (store === null) {
      throw new Error(`Dataset file not found in fallback chain: ${path}`);
    }
    return store.readText(path);
  }

  readJson<T = unknown>(path: string): T {
    return JSON.parse(this.readText(path)) as T;
  }

  describe(): string {
    return `fallback:${this.primary.describe()} -> ${this.fallback.describe()}`;
  }
}

File: python/src/prts_mcp/data/stores.py
from __future__ import annotations

import json
import zipfile
from pathlib import Path
from typing import Any, Protocol


class JsonStore(Protocol):
    """Minimal JSON reader interface for local dataset storage."""

    def exists(self, path: str) -> bool:
        ...

    def read_text(self, path: str) -> str:
        ...

    def read_json(self, path: str) -> Any:
        ...

    def describe(self) -> str:
        ...

    def close(self) -> None:
        ...


def _normalize_path(path: str) -> str:
    normalized = path.replace("\\", "/")
    if normalized.startswith("/"):
        raise ValueError(f"Unsafe dataset path: {path!r}")
    parts = [part for part in normalized.split("/") if part not in ("", ".")]
    if any(part == ".." for part in parts):
        raise ValueError(f"Unsafe dataset path: {path!r}")
    return "/".join(parts)


class DirectoryStore:
    """Read JSON files from a directory root."""

    def __init__(self, root: Path | str) -> None:
        self.root = Path(root)

    def _resolve(self, path: str) -> Path:
        root = self.root.resolve()
        target = (self.root / _normalize_path(path)).resolve()
        if not target.is_relative_to(root):
            raise ValueError(f"Unsafe dataset path: {path!r}")
        return target

    def exists(self, path: str) -> bool:
        return self._resolve(path).is_file()

    def read_text(self, path: str) -> str:
        target = self._resolve(path)
        if not target.is_file():
            raise FileNotFoundError(f"Dataset file not found: {path}")
        return target.read_text(encoding="utf-8")

    def read_json(self, path: str) -> Any:
        return json.loads(self.read_text(path))

    def describe(self) -> str:
        return f"directory:{self.root}"

    def close(self) -> None:
        return None


class ZipStore:
    """Read JSON entries from a zip file."""

    def __init__(self, zip_path: Path | str) -> None:
        self.zip_path = Path(zip_path)
        self._zf: zipfile.ZipFile | None = None

    def _zipfile(self) -> zipfile.ZipFile:
        if self._zf is None:
            self._zf = zipfile.ZipFile(self.zip_path)
        return self._zf

    def close(self) -> None:
        if self._zf is not None:
            self._zf.close()
            self._zf = None

    def __enter__(self) -> ZipStore:
        return self

    def __exit__(self, exc_type: object, exc: object, tb: object) -> None:
        self.close()

    def exists(self, path: str) -> bool:
        inner_path = _normalize_path(path)
        if not self.zip_path.is_file():
            return False
        try:
            self._zipfile().getinfo(inner_path)
            return True
        except KeyError:
            return False

    def read_text(self, path: str) -> str:
        inner_path = _normalize_path(path)
        try:
            with self._zipfile().open(inner_path) as f:
                return f.read().decode("utf-8")
        except KeyError as exc:
            raise FileNotFoundError(f"Dataset zip entry not found: {path}") from exc

    def read_json(self, path: str) -> Any:
        return json.loads(self.read_text(path))

    def describe(self) -> str:
        return f"zip:{self.zip_path}"

    def __del__(self) -> None:
        if getattr(self, "_zf", None) is not None:
            try:
                self.close()
            except Exception:
                pass


class FallbackStore:
    """Read from primary storage first, then fall back to bundled storage."""

    def __init__(self, primary: JsonStore, fallback: JsonStore) -> None:
        self.primary = primary
        self.fallback = fallback

    def _store_for(self, path: str) -> JsonStore | None:
        if self.primary.exists(path):
            return self.primary
        if self.fallback.exists(path):
            return self.fallback
        return None

    def exists(self, path: str) -> bool:
        return self._store_for(path) is not None

    def read_text(self, path: str) -> str:
        store = self._store_for(path)
        if store is None:
            raise FileNotFoundError(f"Dataset file not found in fallback chain: {path}")
        return store.read_text(path)

    def read_json(self, path: str) -> Any:
        return json.loads(self.read_text(path))

    def describe(self) -> str:
        return f"fallback:{self.primary.describe()} -> {self.fallback.describe()}"

    def close(self) -> None:
        self.primary.close()
        self.fallback.close()
